Compile/src/Producers/PUJetIDWeightProducer.cc

```diff
@@ -0,0 +1,92 @@
+#include "Excalibur/Compile/interface/Producers/PUJetIDWeightProducer.h"
+#include "KappaTools/RootTools/interface/Matching.h"
+
+std::string PUJetIDWeightProducer::GetProducerId() const { return "PUJetIDWeightProducer"; }
+
+void PUJetIDWeightProducer::Init(ZJetSettings const& settings) {
+
+    if (settings.GetInputIsData())
+    {
+        LOG(FATAL) << "PUJetID SFs can only be applied to MC!";
+    }
+
+    if (settings.GetPUJetID() != "none")
+    {
+        LOG(INFO) << "Loading PUJetID Efficiencies from file " << settings.GetPUJetIDEffFilename();
+        TFile PUEffFile(settings.GetPUJetIDEffFilename().c_str(), "READONLY");
+        if (PUEffFile.IsOpen() == false)
+        {
+            LOG(FATAL) << "File for PUJetID scalefactors not found. Please check PUEffFilename.";
+        }
+
+        std::string histName = settings.GetPUJetIDEffHistogramName();
+        LOG(INFO) << "Loading histogram " << histName << " for PUJetID scalefactors";
+        LOG(INFO) << "Using PUJetID " << settings.GetPUJetID() << ". Please verify the settings for the efficiency match the used PUJetID.";
+        m_sfhisto = (TH2F*)PUEffFile.Get(histName.c_str());
+        if (!m_sfhisto)
+        {
+            LOG(FATAL)<<"No histogram found with name " << histName;
+        }
+        m_sfhisto->SetDirectory(0);
+
+        std::string errHistName = settings.GetPUJetIDEffErrHistogramName();
+        LOG(INFO) << "Loading histogram " << errHistName << " for PUJetID scalefactors uncertainties.";
+        m_errhisto = (TH2F*)PUEffFile.Get(errHistName.c_str());
+        if (!m_errhisto)
+        {
+            LOG(FATAL)<<"no uncertainty histogram found with name" << errHistName;
+        }
+        m_errhisto->SetDirectory(0);
+
+        PUEffFile.Close();
+    }
+}
+
+void PUJetIDWeightProducer::Produce(ZJetEvent const& event,
+                                    ZJetProduct& product,
+                                    ZJetSettings const& settings) const {
+
+    LOG(DEBUG) << "\n[" << this->GetProducerId() << "]";
+    double sf = 1.0;
+    double err = 0.0;
+
+    const std::vector<std::shared_ptr<KJet>> theJets = product.m_correctedZJets["L1L2L3"];
```

Comment:
Why do you restrict this to L1L2L3 corrected jets? Should we use the fully corrected jets for this?

Reply:
Only data